### PR TITLE
Refactor and test lex API. Simplify standalone use of lex module. Fix couple of bugs. Beautify test code.

### PR DESCRIPTION
**Pull request by Renelvon**
### Lex:
- Add `tokenize` and `quiet_tokenize` for ease of use.
- Factor inner lexer setup in separate method.
- [bugfix] Explicitly bind `logger` of `_LexerFactory` to `logger` of `Lexer`.
- [bugfix] Renew inner lexer on each data input.
- Drop debugging info; `verbose` should serve only one purpose.
- Improve API of `Lexer`. Throw `Exception` when requesting infeasible operations.
- Store `debug`, `optimize` and `verbose` status in `Lexer`.
- Improve docstrings and beautify code.
### Lex test:
- Expand and groom API testing using TDD.
- Cleaning `logger` after faulty input is user's responsibility.
- Split tests in many classes.
- Fix typos and beautify code.

### CODE REVIEW

File: compiler/lex.py [hunk 1]
```diff
@@ -579,8 +569,16 @@ def __next__(self):
...
+    return lexer.tokenize(data)
+
+
+def quiet_tokenize(data):
+    """
+    Lex the given string using the default Lexer.
+    Return an iterator over the string tokens.
+    Explicitly silence errors/warnings.
+    """
+    return tokenize(data, logger=error.LoggerMock())
```
Comment (dionyziz) [inline compiler/lex.py:598]: Why isn't this the default behavior of `tokenize` when `logger` is set to `None`?

File: compiler/lex.py [hunk 2]
```diff
@@ -213,13 +213,15 @@ def __init__(self, logger, verbose=False):
         """
         self.logger = logger
         self.verbose = verbose
-        if self.verbose:
-            self.logger.info(
-                "%s: %s: %s",
-                __name__,
-                self.__class__.__name__,
-                'wrapper initialized'
-            )
```
Comment (dionyziz) [inline compiler/lex.py:222]: Why are we removing these `info` calls?

File: compiler/lex.py [hunk 3]
```diff
@@ -213,13 +213,15 @@ def __init__(self, logger, verbose=False):
...
+
+    def reset(self):
+        """Reset the lexer's state."""
+        self.bol = -1
+        self.level = 0
+        self.logger.clear()
+        self.lexer.begin('INITIAL')
+        self.lexer.lexpos = 0
+        self.lexer.lineno = 1
```
Comment (dionyziz) [inline compiler/lex.py]: Why are these needed?

File: compiler/lex.py [hunk 4]
```diff
@@ -213,13 +213,15 @@ def __init__(self, logger, verbose=False):
...
         self.verbose = verbose
-        if self.verbose:
-            self.logger.info(
-                "%s: %s: %s",
-                __name__,
-                self.__class__.__name__,
-                'wrapper initialized'
-            )
+
+    def reset(self):
```
Comment (dionyziz) [inline compiler/lex.py]: As a public method, this needs testing.

File: compiler/lex.py [hunk 5]
```diff
@@ -279,8 +274,9 @@ def token(self):
         return tok
 
     def input(self, lexdata):
-        """Feed the lexer with input."""
+        """Feed the lexer with input and reset the state."""
         self.lexer.input(lexdata)
+        self.reset()
```
Comment (dionyziz) [inline compiler/lex.py]: Test that `reset` is called.

File: compiler/lex.py (hunk 1 above)
Comment (Renelvon) [inline compiler/lex.py:598]: I believe silencing warnings/errors should never be the default behaviour.

File: compiler/lex.py (hunk 2 above)
Comment (Renelvon) [inline compiler/lex.py:222]: Frankly, they are kinda useless. If something goes wrong, Python's runtime will print an informative stacktrace. Also, `self.verbose` should have one use only.

File: compiler/lex.py (hunk 3 above)
Comment (Renelvon) [inline compiler/lex.py]: Check the tests added in 2a7ac8614415a2c938e50e50a1ddd93b76f97f23 and the commit message of 9ff9154fecd731e0fd7d549f9558534abe3773b3 .

File: compiler/lex.py (hunk 4 above)
Comment (Renelvon) [inline compiler/lex.py]: It's a public method of a private class!

File: compiler/lex.py (hunk 5 above)
Comment (Renelvon) [inline compiler/lex.py]: If `reset` wasn't called there, the tests in https://github.com/dionyziz/llama/commit/2a7ac8614415a2c938e50e50a1ddd93b76f97f23 would fail. Try it yourself. Apart from that, there is no sensible way to test `reset` since it belongs to a private class.

File: compiler/lex.py (hunk 3 above)
Comment (dionyziz) [inline compiler/lex.py]: Why do we have to reuse lexer objects anyway? It's not costly to recreate one.

File: compiler/lex.py (hunk 4 above)
Comment (dionyziz) [inline compiler/lex.py]: OK this is fair.

File: compiler/lex.py (hunk 5 above)
Comment (dionyziz) [inline compiler/lex.py]: You're right, yes.

File: compiler/lex.py (hunk 1 above)
Comment (dionyziz) [inline compiler/lex.py:598]: OK

File: compiler/lex.py (hunk 3 above)
Comment (Renelvon) [inline compiler/lex.py]: Yes, maybe we can do that. I 'll think about it.

Comment (Renelvon) [inline compiler/lex.py]: I redesigned the bugfix based on your suggestion. Have a look, please.
